Bluetooth: Controller: Preliminary support for nRF54L15 SoC

cvinayak opened the pull request:
Add preliminary support for nRF54L15 SoC. This commit does
not support Controller Random Number Generation and
Controller Cryptography (AES-128 encryption) commands, nor
does it support encrypted connections.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Introduce BT_CTLR_CRYPTO_SUPPORT

Introduce BT_CTLR_CRYPTO_SUPPORT so that preliminary port to
support nRF54L15 SoC can be upstreamed without encryption
support.

ENTROPY_GENERATOR now selected when BT_CTLR_CRYPTO enabled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix BT_CTLR_LE_ENC conditional compilation

Fix BT_CTLR_LE_ENC conditional compilation when feature is
disabled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Refactor BT_CTLR_LE_ENC implementation

Refactor reused function in BT_CTLR_LE_ENC feature.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use NRF_RTC and RADIO_SHORTS_TRX_END_DISABLE_Msk

Use NRF_RTC and NRF_RADIO_SHORTS_TRX_END_DISABLE_Msk instead
to prepare towards using configurable use of RTC and Radio
hardware defines.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Preliminary support for nRF54L15 SoC

Add preliminary support for nRF54L15 SoC. This commit does
not support Controller Random Number Generation and
Controller Cryptography (AES-128 encryption) commands, nor
does it support encrypted connections.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix ENTROPY_NRF5_RNG conditional compile

Fix ENTROPY_NRF5_RNG conditional compile when not enabled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c:
```diff
@@ -12,8 +12,12 @@
 #include "hal/debug.h"
 
 #ifndef NRF_RTC
+#if defined(CONFIG_SOC_SERIES_NRF54LX)
```
For simulation all these (in all files) `CONFIG_SOC_SERIES_NRF54LX` would need to be `CONFIG_SOC_COMPATIBLE_NRF54LX`

aescolar commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -15,7 +15,7 @@ config BT_LLL_VENDOR_NORDIC
 
 	select BT_HAS_HCI_VS
-	select BT_CTLR_CRYPTO_SUPPORT
+	select BT_CTLR_CRYPTO_SUPPORT if !SOC_SERIES_NRF54LX
```
`	select BT_CTLR_CRYPTO_SUPPORT if !SOC_COMPATIBLE_NRF54LX`
so it works in simulation too

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1457,6 +1480,10 @@ void radio_tmr_stop(void)
 	nrf_timer_task_trigger(SW_SWITCH_TIMER, NRF_TIMER_TASK_SHUTDOWN);
 #endif /* !CONFIG_BT_CTLR_TIFS_HW */
+
+#if defined(CONFIG_SOC_SERIES_NRF54LX)
```
Except this one (leave as is with the real SOC, as we don't have the POWER registers in the model yet for this target)

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf54lx.h:
```diff
@@ -0,0 +1,639 @@
+
+static inline uint32_t hal_radio_tx_power_floor(int8_t tx_power_lvl)
+{
+	if (tx_power_lvl >= (int8_t)RADIO_TXPOWER_TXPOWER_0dBm) {
```
These values are not anymore 8bits 2's complement representations of the power in dBm, but a funny mapping. So -40dBm (0x130) is bigger than -20dBm (0x1) or even 0dBm (0x13) :(

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -314,12 +322,11 @@ void radio_phy_set(uint8_t phy, uint8_t flags)
+#endif /* !CONFIG_SOC_SERIES_NRF51X && !CONFIG_SOC_SERIES_NRF54LX */
 }
 
 void radio_tx_power_set(int8_t power)
```
327/334: needs special handling for the 54 too, or?

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -314,12 +322,11 @@ void radio_phy_set(uint8_t phy, uint8_t flags)
+#endif /* !CONFIG_SOC_SERIES_NRF51X && !CONFIG_SOC_SERIES_NRF54LX */
 }
 
 void radio_tx_power_set(int8_t power)
```
power will need to be int16_t at least (TXPOWER uses 9 bits) (unless you start by limiting this to values >= -26dBm and hope the mapping does not move out of range for other 54L devices)
Same applies to radio_tx_power_max_set/radio_tx_power_min_get / radio_tx_power_floor()
and `lll_radio_tx_pwr_floor()` and lll_test:tx_power_set() and wherever this propagates to..

cvinayak replied:
Updated

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -11,18 +11,21 @@ config BT_LLL_VENDOR_NORDIC
 
 	select BT_HAS_HCI_VS
-	select BT_CTLR_LE_ENC_SUPPORT if !BT_CTLR_DATA_LENGTH_CLEAR && \
+	select BT_CTLR_CRYPTO_SUPPORT if !SOC_SERIES_NRF54LX
```
@wopu-ot or @thoh-ot  This PR introduces explicit need to select `BT_CTLR_CRYPTO_SUPPORT` in out-of-tree LLL variants.

cvinayak committed:
boards: nordic: nrf54l15pdk: Bind Bluetooth C2H and monitor uart

Bind Bluetooth Controller to Host and Monitor UART.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: hci_uart: Add preliminary support for nrf54l15pdk

Add preliminary support for building hci_uart sample
for nrf54l15pdk.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

aescolar committed:
Bluetooth: Controller: Treat nrf54l15bsim like a real platform

Use the compatible kconfig option so that for the simulated
nRF54L15 we build the same code as for the real platform.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
Bluetooth: Controller: Use HAL to modify renamed registers in nRF54

In some nRF54 devices, DATAWHITEIV was renamed to
DATAWHITE,
and the CRCCNF SKIADDR field was renamed OFFSET.
The nrf HAL hid this change internally,
so let's use it so we don't need to ifdef these
in the Bluetooth Controller HAL code.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -15,7 +15,7 @@ config BT_LLL_VENDOR_NORDIC
 
 	select BT_HAS_HCI_VS
-	select BT_CTLR_CRYPTO_SUPPORT if !SOC_SERIES_NRF54LX
+	select BT_CTLR_CRYPTO_SUPPORT if !SOC_COMPATIBLE_NRF54LX
```
@aescolar `SOC_COMPATIBLE_NRF54LX` is not selected for `SOC_SERIES_NRF54LX` causing build failure.

cvinayak committed:
samples: Bluetooth: hci_uart: Use lower IRQ priority SoC peripherals

Use lower IRQ proirities for SoC peripherals, example UART,
to reduce ISR latencies on the IRQs processed by the
Controller.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

aescolar committed:
Bluetooth: Controller: Fix BT_CTLR_DATA_LEN_UPDATE_SUPPORT selection

For the Nordic HW, the BT_CTLR_DATA_LEN_UPDATE_SUPPORT
does not require CCM HW enabled, hence support Data Length
Update if Encryption Support is disabled.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

aescolar reviewed (dismissed):
Thank you :)
Having support (even if without encryption) for the 54 will enable us to test it much better in CI, and should avoid quite many regressions

Thalley commented on samples/bluetooth/hci_uart/sample.yaml:
```diff
@@ -58,3 +58,13 @@ tests:
     tags:
       - uart
       - bluetooth
+  sample.bluetooth.hci_uart.nrf54l15.all:
+    harness: bluetooth
+    platform_allow: nrf54l15pdk/nrf54l15/cpuapp
+    integration_platforms:
+      - nrf54l15pdk/nrf54l15/cpuapp
+    extra_args:
+      - OVERLAY_CONFIG=overlay-all-bt_ll_sw_split.conf
+    tags:
+      - uart
+      - bluetooth
```
This looks odd to me. For the nRF53 we only add the netcore here, but this is for the appcore. 
Similarly, we do not use an overlay for the nRF53, but we do here.

Thalley commented on subsys/bluetooth/common/rpa.c:
```diff
@@ -23,34 +23,33 @@
 
-#if defined(CONFIG_BT_CTLR) && defined(CONFIG_BT_HOST_CRYPTO)
+#if defined(CONFIG_BT_CTLR_CRYPTO) && defined(CONFIG_BT_HOST_CRYPTO)
 #include "../controller/util/util.h"
 #include "../controller/hal/ecb.h"
-#endif /* defined(CONFIG_BT_CTLR) && defined(CONFIG_BT_HOST_CRYPTO) */
+#endif /* CONFIG_BT_CTLR_CRYPTO && CONFIG_BT_HOST_CRYPTO */
```
Why do we need to enable both host and controller crypto? Should this have been a ` || `?

Thalley replied:
Ditto for the below checks

Thalley commented on subsys/bluetooth/controller/CMakeLists.txt:
```diff
@@ -173,10 +173,6 @@ zephyr_library_sources_ifdef(
 zephyr_library_include_directories(
   .
   include
-  )
-
-zephyr_library_include_directories_ifdef(
-  CONFIG_BT_CTLR_CRYPTO
```
The commit for this change just says "fix" without any additional information. 

Why are now always including crypto, even if it is disabled?

cvinayak replied:
as `IS_ENABLED` is used in the other places that have calls to Host crypto interface, it makes it necessary to always include the prototypes in such files.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:
```diff
@@ -319,22 +323,38 @@ int lll_deinit(void)
 
 int lll_csrand_get(void *buf, size_t len)
 {
+#if defined(CONFIG_ENTROPY_NRF5_RNG)
 	return entropy_get_entropy(dev_entropy, buf, len);
+#else
+	return 0;
+#endif
```
For this and the other functions in this commit, shouldn't we return `-ENOTSUP` instead?

As per the https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html if a function is not implemented or supported, it should return that error. 

Returning `0` when trying to get a value seems like a bad idea.

cvinayak replied:
Added comment.

Thalley commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -801,9 +801,14 @@ config BT_CTLR_SYNC_ISO
 
+config BT_CTLR_BROADCAST_ISO_ENC
+	bool
+	depends on BT_CTLR_CRYPTO_SUPPORT && BT_CTLR_BROADCAST_ISO
+	select BT_CRYPTO if BT_LL_SW_SPLIT
```
It doesn't depend on `BT_CRYPTO`?

Should it have depended on `BT_CRYPTO` instead of `BT_CTLR_CRYPTO_SUPPORT` rather than `select`ing `BT_CRYPTO`?

cvinayak replied:
`BT_CRYPTO` is a module used by Host or Controller or both; and is a hidden Kconfig that needs to be selected by either Host or Controller.

Now, looking deep, maybe TinyCrypt is getting included, which is not needed by the Controller.... i will need to follow up in a separate PR.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5.h:
```diff
@@ -44,6 +47,14 @@
+#include <hal/nrf_rtc.h>
+#include <hal/nrf_timer.h>
+
+#if defined(CONFIG_BT_CTLR_LE_ENC) || defined(CONFIG_BT_CTLR_BROADCAST_ISO_ENC)
```
Since `defined(CONFIG_BT_CTLR_LE_ENC) || defined(CONFIG_BT_CTLR_BROADCAST_ISO_ENC)` appears so many places, we could consider having a hidden helper Kconfig option to check against instead

cvinayak replied:
any suggested name?

Thalley replied:
Nothing that comes to mind that wouldn't get confused with other values :)

Thalley commented on samples/bluetooth/beacon/sample.yaml:
```diff
@@ -6,13 +6,20 @@ tests:
     integration_platforms:
       - qemu_cortex_m3
-
+      - nrf51dk/nrf51822
+      - nrf52dk/nrf52832
+      - nrf54l15pdk/nrf54l15/cpuapp
```
Seems odd to add the nRF 51, 52 and 54 series without the nRF 53 series.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf54lx.h:
Not sure what is common for the hal files, but there are **a lot** of magic numbers in this file.

cvinayak replied:
Most of the values are silicon specifics. For now, i have just used values from other SoC variant, but truely need to be specific value that h/w verification team should provide.

Thalley replied:
Right, but something like `BIT(0)` for the PHY could be defined in the file, so you can do `case PHY_1M` instead of `case BIT(0)` for increased readability and easier maintenance. 

My personal rule of thumb is that if any literal value is used more than once, then define it

cvinayak replied:
Need refactoring if include headers to be able to use `PHY_1M` and similar defines. Will leave it to another PR.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf54lx.h:
```diff
@@ -0,0 +1,795 @@
+	return mode;
+}
+
+static inline int8_t hal_radio_tx_power_max_get(void)
+{
+	return 8; /* +8 dBm */
+}
+
+static inline int8_t hal_radio_tx_power_min_get(void)
+{
+	return -46; /* -46 dBm */
+}
```
Does it make sense to use `inline` for these type of functions?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf54lx.h:
```diff
@@ -0,0 +1,795 @@
+	return -46;
+}
+
+static inline uint32_t hal_radio_tx_power_value(int8_t tx_power_lvl)
```
Did you consider lookup tables for this and `hal_radio_tx_power_floor`?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:
```diff
@@ -319,22 +323,54 @@ int lll_deinit(void)
 	return entropy_get_entropy(dev_entropy, buf, len);
+#else
+	/* FIXME: No suitable entropy device available yet.
+	 *        It is required by Controller to use random numbers.
```
Is this a requirement from the BT spec? 

If so, then uninitialized values are probably not a valid random value. 

When is this an issue? Do we not have entropy support for the nRF54L?

Thalley replied:
I don't understand why it shouldn't return an error then. 

We are just returning uninitialized values and err == 0, which not only would be visible in e.g. valgrind, but also not complying with any potential randomness requirements. 

This does not look good to me, but I see that both @carlescufi and @aescolar have approved it, so maybe it's not a real issue.

cvinayak replied:
The scope of this PR is limited to preliminary support and stated in the commit as:

```
Bluetooth: Controller: Preliminary support for nRF54L15 SoC

Add preliminary support for nRF54L15 SoC. This commit does
not support Controller Random Number Generation and
Controller Cryptography (AES-128 encryption) commands, nor
does it support encrypted connections.
```

The board is also a Preview Dev Kit, hence, there will be updates to come in the future.

Thalley replied:
Fair, had missed that. Thanks!

cvinayak committed:
Bluetooth: Controller: Use BT_HCI_ERR_UNSPECIFIED as needed

A Host shall consider any error code that it does not
explicitly understand equivalent to the error code
Unspecified Error (0x1F).

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add explicit LLCP error code check

Add unit tests to cover explicit LLCP error code check and
cover the same in the Controller implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley reviewed (dismissed):
Looks OK to me

thoh-ot reviewed (approved):
Only looked at the LLCP changes.

erbr-ot reviewed (approved):
LLCP changes reviewed
